lib/puppet/provider/mount/parsed.rb

```diff
@@ -272,6 +272,49 @@ def self.mountinstances
     instances
   end
 
+  # Sort fstab entries so that
+  # a mount point is located inside another mount point appears later and
+  # a device that is located inside a mount point appears later.
+  #
+  # "Stable opportunistic insertion sort"
+  # Copy unsorted list A to sorted B by repeatedly
+  # * removing the first element from A
+  # * compare with each element in B, from first to last
+  # * insert before current element in B if smaller
+  # * insert at the end of B otherwise
+  # * finish if A is empty
+  # * repeat
+  #
+  # Stabilizes itself by inserting as late as possible.
+  #
+  # A record is "smaller" than another if
+  # * the mount point of A is a leading substring of B's mount point
+  # * the mount point of A is a leading substring of B's device (B is a bind mount)
+  #
+  # @note Does not really care about directory names and does substring
+  #   comparison only. So /media will be put before /mediastuff. This is
+  #   not necessary but will be rare and usually cause no harm, either.
+  #
+  # @api private
+  def self.order(records)
```

Comment:
Rather than an O(n^2) sort with a nested catch/throw (I would imagine that the Ruby VM as it needs to setup an exception handler in the frame, which is an expensive operation on some platforms), would this work?

```ruby
def self.order(records)
  # Create an array of indexes into the given records array
  indexes = [*0..(records.size - 1)]
  indexes.sort! do |a, b|
    # First sort by mount point
    result = records[a][:name] <=> records[b][:name]
    # If b's mountpoint should come before a's, sort by device
    result = records[a][:device] <=> records[b][:device] if result >= 0
    # If equal, sort by index to create a stable sort
    result = a <=> b if result == 0
    result
  end.map! do |index|
    # Replace the index with the original entry
    records[index]
  end
end
```